### ISSUE TITLE
classmethod or member method for create function in Python API

### ISSUE THREAD

**Issue by AlbertDeFusco** [conda-incubator/conda-project#36]
For reference: https://github.com/conda-incubator/conda-project/pull/22#discussion_r956135892

right now to create a new project you can

```python
from conda_project import CondaProject

new_project = CondaProject.create('project-directory', dependencies=['python=3.8'])
```

This was initially modeled after Pandas classmethods like [from_dict](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.from_dict.html).

An alternative proposed by @mattkram is

```python
from conda_project import CondaProject

new_project = CondaProject('project-directory')
new_project.create(dependencies=['python=3.8'])
```

**Comment by mattkram** [comment]
I'd like to outline my thought process that led to those comments, and I am totally open to all discussion and feedback.

To me, I've personally seen the `@classmethod`-based constructors in the form of `.from_something()`, as you mentioned from `pandas`, and this pattern is very common in other languages too. But the action of `.create()` feels like it is specific to a single `CondaProject` instance, which it is.

So although the behavior is the same whether we use a `@classmethod` or an instance method, thinking forward to the things I would like to do *with* a specific project instance, I think it would be more clear to have an instance of that type, and then perform actions on it, e.g.

```python
project = CondaProject('project-directory')
project.create(dependencies=['python=3.8'])
project.lock()
project.run("command_name") # Making up actions
```

In this way, the `CondaProject` instance establishes the common bit (analogous to being within a certain directory), and any actions you may take on that project are instance methods (analogous to CLI commands).

I would recommend we do this, because it is more consistent. And if we wanted, we could add a `@classmethod` like the following for an alternative constructor if it is worth it for a more clear API from other perspectives:

```python
project = CondaProject.from_cli_args(..., create=True)
...
```

The above instance-method-based API would also unlock a cool context-manager API if we wanted:

```python
with CondaProject('project-directory') as project:
    project.create(dependencies=['python=3.8'])
    project.lock()
    project.run("command_name") # Making up actions
```

And I also personally like method chaining (I know some people don't), which if we return `self` from each method we could do:

```python
with CondaProject('project-directory') as project:
    (
        project
        .create(dependencies=['python=3.8'])
        .lock()
        .run("command_name") # Making up actions
    )
```

which, among other things would surely lead to some lively discussions about readability!

**Comment by mattkram** [comment]
> Ah, I like these ideas. I'm curious if you have a use case for the context manager that we might write up into the docs as well. Would this create `project-directory` if it doesn't exist and trash it after the command is run?

I was thinking more of just "run these commands in this working directory", but potentially there could be a `destroy_on_exit=True` or something to do a `rm -rf` of the project directory. I guess this could be a cool way of running some command in an ethereal way, but I'm not convinced that use case is that strong (yet).

I was more thinking it would be analagous to doing this, just allowing running `CondaProject` methods inside a specific directory:

```python
import os
from contextlib import contextmanager

@contextmanager
def new_cwd(path):
    old_cwd = os.getcwd()
    os.chdir(path)
    yield
    os.chdir(old_cwd)

with new_cwd('/some/path'):
    # do some stuff
```

If we did that, we may also want to have an option or default to create the directory if it doesn't exist (and optionally clean up afterwards?

**Comment by mattkram** [comment]
Although, on second thought, this may not be necessary if the methods themselves handle the directory changes ...

**Comment by AlbertDeFusco** [comment]
Yes, each instance of CondaProject is self-contained for its directory with resolved paths.

So, if these two directories already exist and have appropriate yaml files I can operate them independently. Like git "project" and "directory" are really the same thing.

```python
p1 = CondaProject('project-dir-1')
p2 = CondaProject('../project-dir-2')

p1.default_environment.lock()
p1.default_command.run()

p2.environment['dev'].prepare()
```

As of #22 there is no top-level `CondaProject.lock()` or `CondaProject.prepare` method, but they can be added to operate like the CLI.

`CondaProject.__init__` resolves the path provided (assumes `.` if not supplied)

https://github.com/conda-incubator/conda-project/blob/0f7806d55d754b86c97de2e7cf038abb9cd2421f/conda_project/project.py#L343-L344

Now that I think about it there may be good uses of a context manager that does `os.chdir(project.directory)`. For example you could create a project factory that can build archives or docker images. (this is not really well thought out, but something that I think is reasonable)

```python
with CondaProject('temporary-directory', delete_on_exit=True) as project:
    (project
      .create(dependencies=..., conda_configs=..., platforms=..., commands={...})
      .lock()
      .prepare()
    )

    # copy in some files like scripts, notebooks, data
    # do other file manipulations that you want to archive along with the 
    # prepared environment in a conda-pack archive inside the final project archive
    
    project.archive(pack_envs=True, output_path='<some-more-permanent-location>.tar.bz2')
```

**Comment by bkreider** [comment]
I like the instance method (__init__), but this conversation is confusing until it's defined what it means to have a conda_project object. 

Can you have a conda_project object without writing to disk?  It sounds weird, but there might be cases where you want to load and think about a conda-project without materializing it / serializing anything to disk. 

Say you want to store conda-projects in a DB and then load them into memory and decide later which one you want to "materialize" and run on disk.  Is a minimal conda-project bits in a directory or just some metadata?  (definition and optionally lock-data).  Even just loading a number of conda-projects into memory to manipulate or compare them.
